pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go

```diff
@@ -638,13 +647,78 @@ func (a *index) inRevision(obj any) bool {
 
 // All return all known workloads. Result is un-ordered
 func (a *index) All() []model.AddressInfo {
-	res := slices.Map(a.workloads.List(), modelWorkloadToAddressInfo)
+	var res []model.AddressInfo
+	// Add all workloads
+	for _, wl := range a.workloads.List() {
+		// If EnableAmbientMultiNetwork is enabled, we only want to add workloads that are local to this cluster
+		// Non cluster local workloads will be added by the service lookup
+		if features.EnableAmbientMultiNetwork && wl.Workload.ClusterId != a.ClusterID.String() {
+			continue
+		}
+		res = append(res, wl.AsAddress)
+	}
+	// Add all services
+	// TODO: Update to split horizon
 	for _, s := range a.services.List() {
+		if features.EnableAmbientMultiNetwork && s.Scope == model.Global {
+			// If EnableAmbientMultiNetwork is enabled, we want to add workloads that correspond to global services
+			for _, wl := range a.workloads.ByServiceKey.Lookup(s.ResourceName()) {
+				if wl.Workload.ClusterId != string(a.ClusterID) {
```

Comment:
@keithmattix @Stevenjin8 I thought the logic in `All` would be changing more for split horizon - if not, I can remove the TODO

Reply:
Ok cool, yeah let's tackle it in your OPR

Reply:
right I did say that. As I refactored it, I think it should be feature flag agnostic now.